spec/ics-027-interchain-account/README.md

```diff
@@ -0,0 +1,283 @@
+}
+```
+```typescript
+function onChanOpenAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  version: string) {
+  // version not used at present
+  abortTransactionUnless(version === "")
+  // port has already been validated
+}
+```
+```typescript
+function onChanOpenConfirm(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier) {
+  // accept channel confirmations, port has already been validated
+}
+```
+```typescript
+function onChanCloseInit(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier) {
+  // no action necessary
+}
+```
+```typescript
+function onChanCloseConfirm(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier) {
+  // no action necessary
+}
+```
+### **Packet relay**
+
+In plain English, between chains `A` and `B`. It will describe only the case that chain A wants to register an Interchain account on chain B and control it. Moreover, this system can also be applied the other way around. In these set of examples, we follow Cosmos-SDK's modularity rule and its `msg` system.
+```typescript
+interface ChainAccountTx {
+  msgs: Msg[]
+}
+```
+```typescript    
+function onSendPacket(packet: Packet) {  
+  // Sending chain can do preliminary work (i.e. create a log of transactions sent) if needed.
+}
+```
+```typescript
+function onRecvPacket(packet: Packet): bytes {
+  if (packet.data is RunTxPacketData) {
+    ChainAccountTx data = {}
+    // Decode transaction request
+    codec.unmarshal(packet.data.txBytes, &data)
+      
+    signers = []
+    for (const msg of data.msgs) {
+      signers.push(msg.getSigners())  
+    }
+      
+    path = "{packet.sourcePort}/{packet.sourceChannel}"
+      
+    // Check tx has right permissions
+    for (const signer of signers){
+      abortTransactionUnless(path === addressesRegisteredChannel[signer])
+    }
+    
+    // Get handler from router
+    handler = router(msg.route())
+    abortTransactionUnless(handler != null)
+        
+    // Execute handler with msg
+    result = handler(msg)
+        
+    if (result.code === 0) {
+        // Return acknowledgement byte as 0x0 if tx succeeds.
+        return 0x0
+    } else {
+        // Return acknowledgement byte as error code if tx fails.
+        return binary.littleEndian.encode(result.code)
+    }
+  }
+    
+  if (packet.data is RegisterIBCAccountPacketData) {
+    RegisterIBCAccountPacketData data = packet.data
+      
+    path = "{packet.sourcePort}/{packet.sourceChannel}"
+    address = sha256(path + packet.salt)
+      
+    // Should not block even if there is normal account,
+    // because attackers can distrupt to create an ibc managed account
+    // by sending some assets to estimated address in advance.
+    // And IBC managed account has no public key, but its sequence is 1.
+    // It can be mark for Interchain account, becuase normal account can't be sequence 1 without publish public key.
+    account = accountKeeper.getAccount(account)
```

Comment:
Also `accountKeeper` seems Cosmos specific. As an alternative you declare an interface named `accountKeeper` with a set of functions with some constraints(which will be inspired from Cosmos' `accountKeeper` design, but works chain agnostic)

Reply:
Here is where I would like to see a `runTx` function.